executor.go

```diff
@@ -0,0 +1,227 @@
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	tea "github.com/charmbracelet/bubbletea"
+	"github.com/minio/confess/tests"
+	"github.com/minio/confess/utils"
+	"github.com/minio/minio-go/v7"
+	"github.com/minio/pkg/console"
+	"github.com/minio/pkg/ellipses"
+)
+
+var slashSeparator = "/"
+
+// Signature denotes the s3 signature type.
+type Signature string
+
+const (
+	// SignatureV4 denotes s3v4 algorithm
+	SignatureV4 Signature = "s3v4"
+	// SignatureV2 denotes s3v2 algorithm
+	SignatureV2 Signature = "s3v2"
+)
+
+// Config represents the confess executor configuration.
+type Config struct {
+	Hosts      []string      `json:"hosts"`
+	AccessKey  string        `json:"accessKey"`
+	SecretKey  string        `json:"secretKey"`
+	Insecure   bool          `json:"insecure,omitempty"`
+	Region     string        `json:"region,omitempty"`
+	Signature  Signature     `json:"signature"`
+	Bucket     string        `json:"bucket"`
+	OutputFile string        `json:"outputFile"`
+	FailAfter  int64         `json:"failAfter"`
+	Duration   time.Duration `json:"duration"`
+}
+
+// Test interface defines the test.
+type Test interface {
+	Name() string
+	Init(ctx context.Context, config tests.Config, stats *tests.Stats) error
+	Setup(ctx context.Context) error
+	Run(ctx context.Context) error
+	TearDown(ctx context.Context) error
+}
+
+// Validate - validates the config provided.
+func (c Config) Validate() error {
+	if len(c.Hosts) == 0 {
+		return errors.New("empty hosts")
+	}
+	if c.AccessKey == "" || c.SecretKey == "" {
+		return errors.New("accessKey and secretKey must be set")
+	}
+	if c.Bucket == "" {
+		return errors.New("bucket name should not be empty")
+	}
+	return nil
+}
+
+// Executor represents the test executor.
+type Executor struct {
+	Bucket               string
+	Clients              []*minio.Client
+	EnableVersionedTests bool
+	LogFile              string
+	Stats                *tests.Stats
+	Duration             time.Duration
+	FailAfter            int64
+	sLock                sync.Mutex
+}
+
+// NewExecutor returns the instance of the type Executor.
+func NewExecutor(ctx context.Context, config Config) (*Executor, error) {
+	if err := config.Validate(); err != nil {
+		return nil, err
+	}
+	var endpoints []string
+	for _, hostStr := range config.Hosts {
+		hosts := strings.Split(hostStr, ",")
+		for _, host := range hosts {
+			if len(host) == 0 {
+				continue
+			}
+			if !ellipses.HasEllipses(host) {
+				endpoints = append(endpoints, host)
+				continue
+			}
+			patterns, perr := ellipses.FindEllipsesPatterns(host)
+			if perr != nil {
+				console.Fatalln(fmt.Errorf("unable to parse input arg %s: %s", patterns, perr))
+			}
+			for _, values := range patterns.Expand() {
+				endpoints = append(endpoints, strings.Join(values, ""))
```

Comment:
Why do we want to append strings together. i.e. ["aa", "bb", "cc"] becomes "aabbcc"?

Reply:
as per `minio/pkg/ellipses` - only one string will be there in inner list

```
{1...64} will expand to `[][]string{{"1"}, {"2"}, {"3"}, {"4"}...{"64"}}`

Reply:
I don't understand. `strings.Join()` joins slices by separator. In what scenario you would need to join strings together?

Reply:
here the  `len(values)` will be `1`

so, the strings.Join is just for converting []string to string @balamurugana

Reply:
ok, let me check few examples from that pkg and change this